File: src/main/java/com/cpj/swagger/APIParser.java
/*
 *
 * Copyright (c) 2011, 2016 CPJ and/or its affiliates. All rights reserved.
 * 
 */
package com.cpj.swagger;

import java.io.File;
import java.io.FileWriter;
import java.io.IOException;
import java.lang.reflect.Method;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Properties;

import org.apache.log4j.Logger;

import com.alibaba.fastjson.JSONWriter;
import com.cpj.common.reflect.ReflectUtils;
import com.cpj.common.util.TextUtil;
import com.cpj.swagger.annotation.API;
import com.cpj.swagger.annotation.APISchema;
import com.cpj.swagger.annotation.APISchemaPropertie;
import com.cpj.swagger.annotation.APISchemas;
import com.cpj.swagger.annotation.APITag;
import com.cpj.swagger.annotation.APITags;
import com.cpj.swagger.annotation.APIs;
import com.cpj.swagger.annotation.Item;
import com.cpj.swagger.annotation.Items;
import com.cpj.swagger.annotation.Param;

/**
 * 接口解析器。
 * 
 * <p>可以通过如下的方式来构建一个接口解析器：</p>
 * 
 * <pre>
 * 	// 创建一个构建器
 * 	String host = "127.0.0.1/app";
 * 	String file = "c:/apis.json";
 * 	String[] packageToScan = new String[]{"com.cpj.demo.api"};
 * 	APIParser.Builder builder = new APIParser.Builder(host, file, packageToScan);
 * 
 *	// 设置可选参数
 * 	builder.basePath("/");
 * 
 * 	// 构建解析器
 * 	APIParser parser = builder.build();
 * 	// 解析
 * 	parser.parse();
 * </pre>
 * <p>或者通过这种方式来构建一个接口解析器：</p>
 * <pre>
 * 	APIParser.newInstance(props);
 * </pre>
 * @author yonghaun
 * @since 1.0.0
 */
public final class APIParser implements APIParseable {
	private final static Logger sLogger = Logger.getLogger(APIParser.class);

	/**
	 * 创建一个解析器。
	 * @param props properties。
	 * @throws IOException
	 * @see APIParser.Builder
	 */
	public final static APIParser newInstance(Properties props) throws IOException {
		String[] packageToScan = props.getProperty("packageToScan").split(";");
		Builder builder = new Builder(props.getProperty("apiHost"), props.getProperty("apiFile"), packageToScan)
										.basePath(props.getProperty("apiBasePath"))
										.description(props.getProperty("apiDescription"))
										.termsOfService(props.getProperty("termsOfService"))
										.title(props.getProperty("apiTitle"))
										.version(props.getProperty("apiVersion"))
										.suffix(props.getProperty("suffix"));
		return new APIParser(builder);
	}
	
	private APIParser(Builder builder) {
		// 扫描class并生成文件所需要的参数
		this.host = builder.host;
		this.file = builder.file;
		this.packageToScan = builder.packageToScan;
		try {
			packages = ReflectUtils.scanPackages(this.packageToScan, true);
		} catch (Exception e) {
			throw new IllegalStateException("扫描包信息失败", e);
		}
		this.basePath = builder.basePath;
		this.suffix = builder.suffix;

		// API文档信息
		info = new APIDocInfo.Builder()
				.contact(builder.contact)
				.description(builder.description)
				.license(builder.license)
				.termsOfService(builder.termsOfService)
				.title(builder.title)
				.version(builder.version).build();
	}

	private String host;
	private String basePath;
	private String suffix = "";
	/**
	 * @author yonghuan
	 */
	public static class Builder {
		// required args
		private String host;
		private String file;
		private List<String> packageToScan;

		/**
		 * 创建一个构建器。
		 * 
		 * @param host
		 *            API访问地址（不包含协议）
		 * @param file
		 *            解析产生的文件的存放路径
		 * @param packageToScan
		 *            待扫描的包
		 */
		public Builder(String host, String file, String[] packageToScan) {
			this(host, file, Arrays.asList(packageToScan));
		}

		/**
		 * 创建一个构建器。
		 * 
		 * @param host
		 *            API访问地址（不包含协议）
		 * @param file
		 *            解析产生的文件的存放路径
		 * @param packageToScan
		 *            待扫描的包
		 */
		public Builder(String host, String file, List<String> packageToScan) {
			this.host = host;
			this.file = file;
			this.packageToScan = packageToScan;
		}

		/**
		 * 构建解析器。
		 * 
		 * @return
		 */
		public APIParser build() {
			return new APIParser(this);
		}

		private String basePath;

		/**
		 * 设置API相对于host（API访问地址）的基路径
		 * 
		 * @param val
		 *            API相对于host（API访问地址）的基路径
		 * @return
		 */
		public Builder basePath(String val) {
			this.basePath = val;
			return this;
		}

		private String suffix = "";
		/**
		 * 设置请求地址的后缀，如：.do、.action。
		 * @param suffix 请求地址的后缀
		 * @return
		 */
		public Builder suffix(String suffix) {
			if(TextUtil.isNotEmpty(suffix)) {
				this.suffix = suffix;
			}
			return this;
		}
		
		private String description;

		/**
		 * 设置API描述
		 * @param val
		 *            API描述
		 * @return
		 */
		public Builder description(String val) {
			this.description = val;
			return this;
		}

		private String version;

		/**
		 * 设置API版本
		 * 
		 * @param val
		 *            API版本
		 * @return
		 */
		public Builder version(String val) {
			this.version = val;
			return this;
		}

		private String title;

		/**
		 * 设置API标题
		 * 
		 * @param val
		 *            API标题
		 * @return
		 */
		public Builder title(String val) {
			this.title = val;
			return this;
		}

		private String termsOfService;

		/**
		 * 设置API开发团队的服务地址
		 * 
		 * @param val
		 *            API开发团队的服务地址
		 * @return
		 */
		public Builder termsOfService(String val) {
			this.termsOfService = val;
			return this;
		}

		private String contact;

		/**
		 * 设置API开发团队的联系人
		 * 
		 * @param val
		 *            API开发团队的联系人
		 * @return
		 */
		public Builder contact(String val) {
			this.contact = val;
			return this;
		}

		private License license;

		/**
		 * 设置API遵循的协议（如apahce开源协议）
		 * 
		 * @param val
		 *            API遵循的协议（如apahce开源协议）
		 * @return
		 */
		public Builder license(License val) {
			try {
				this.license = val.clone();
			} catch (CloneNotSupportedException e) {
				throw new AssertionError(e);
			}
			return this;
		}
	}

	private APIDocInfo info;

	private String file;

	/**
	 * @return 解析完成后存放JSON数据的文件路径。
	 */
	public String getFile() {
		return file;
	}

	private List<String> packageToScan;

	/**
	 * @return 待解析接口所在包
	 */
	public List<String> getPackageToScan() {
		return packageToScan;
	}

	private List<Package> packages;

	private Map<String, Item> items;

	@Override
	public void parse() throws Exception {
		/* 将结果写入文件 */
		File f = new File(file);
		sLogger.debug("生成的文件保存在=>" + f.getAbsolutePath());
		JSONWriter writer = new JSONWriter(new FileWriter(f));
		APIDoc api = (APIDoc) parseAndNotStore();
		writer.writeObject(api);
		writer.flush();
		writer.close();
	}

	@Override
	public Object parseAndNotStore() throws Exception {
		APIDoc api = new APIDoc();
		String[] schemes = new String[] { "http" };
		api.setSchemes(schemes);
		api.setHost(host);
		api.setBasePath(basePath);
		api.setInfo(info);

		/* 解析全部item */
		items = parseItem();

		/* 解析全部tag */
		List<Tag> tags = parseTag();
		api.setTags(tags);

		/* 解析全部path */
		Map<String, Map<String, Path>> paths = parsePath();
		api.setPaths(paths);

		/* 解析全部definition */
		Map<String, Object> definitions = parseDefinition();
		api.setDefinitions(definitions);
		return api;
	}
	
	private Map<String, Item> parseItem() throws Exception {
		Map<String, Item> items = new HashMap<String, Item>();

		for (Package pk : packages) {
			Items items2 = pk.getAnnotation(Items.class);
			if (items2 == null) {
				continue;
			}
			Item[] is = items2.items();
			for (Item i : is) {
				items.put(i.value(), i);
			}
		}
		return items;
	}

	
	/**
	 * url -> [ path ]
	 * 
	 * @return
	 * @throws Exception
	 */
	private Map<String, Map<String, Path>> parsePath() throws Exception,
			IllegalArgumentException {
		Map<String, Map<String, Path>> paths = new HashMap<String, Map<String, Path>>();
		List<Class<?>> clazzs = ReflectUtils.scanClazzs(packageToScan, true); // 扫描包以获取包中的类
		for(Class<?> clazz : clazzs) {
			APIs apis = clazz.getAnnotation(APIs.class);
			if(apis == null) {
				continue;
			}
			List<Method> apiMethods = scanAPIMethod(clazz);
			for (Method method : apiMethods) {
				API service = method.getAnnotation(API.class);
				boolean isMultipart = hasMultipart(service);
				String url = apis.value() + "/" + service.value() + suffix;
				if(url.endsWith("/")) {
					url = url.substring(0, url.length() - 1);
				}
				Map<String, Path> path = paths.get(url); // get/psot/put/delete
				if (path == null) {
					path = new HashMap<String, Path>();
					paths.put(url, path);
				}
	
				Path p = path.get(service.method());
				if (p == null) {
					p = new Path();
					path.put(service.method().toLowerCase(), p);
				}
				if (TextUtil.isNotEmpty(service.description())) {
					p.setDescription(service.description());
				} else {
					p.setDescription(service.summary());
				}
				if (TextUtil.isNotEmpty(service.operationId())) {
					p.setOperationId(service.operationId());
				} else { // 未设置operationId，
					p.setOperationId(method.getName());
				}
				List<String> tags = Arrays.asList(service.tags());
				if(service.tags().length == 0) {
					String ns = apis.value();
					if(ns.startsWith("/")) {
						ns = ns.substring(1);
					}
					tags = Arrays.asList(ns);
				}
				p.setTags(tags);
				p.setSummary(service.summary());
				if (isMultipart) { // multipart/form-data
					p.setConsumes(Arrays.asList("multipart/form-data"));
				} else {
					p.setConsumes(Arrays.asList(service.consumes()));
				}
				p.setProduces(Arrays.asList(service.produces()));
				List<Map<String, Object>> parameters = new ArrayList<Map<String, Object>>(); // 请求参数
	
				/** 解析参数，优先使用schema */
				for (Param rp : service.parameters()) {
					Map<String, Object> parameter = new HashMap<String, Object>();
					if (rp.schema() != null && !rp.schema().trim().equals("")) { // 处理复杂类型的参数
						if (isMultipart) { // 当请求的Content-Type为multipart/form-data将忽略复杂类型的参数
							throw new IllegalArgumentException(
									"请求的Content-Type为multipart/form-data，将忽略复杂类型的请求参数[ "
											+ rp.schema() + " ]");
						}
						parameter.put("in", "body");
						parameter.put("name", "body");
						Map<String, Object> $ = new HashMap<String, Object>();
						$.put("$ref", "#/definitions/" + rp.schema());
						parameter.put("schema", $);
					} else { // 简单类型的参数
						if (isMultipart && !"path".equals(rp.in()) && !"header".equals(rp.in())) { // 包含文件上传
							parameter.put("in", "formData");
							parameter.put("type", rp.type());
						} else { // 不包含文件上传
							String in = rp.in();
							if(TextUtil.isEmpty(in)) {
								if("post".equalsIgnoreCase(service.method())) {
									in = "formData";
								} else {
									in = "query";
								}
							}
							parameter.put("in", in);
							parameter.put("type", rp.type());
							String format = rp.format();
							if(TextUtil.isNotEmpty(format)) {
								parameter.put("format", format);
							}
						}
						parameter.put("name", rp.name());
						parameter.put("description", rp.description());
						parameter.put("required", rp.required());
						if (rp.items() != null && !rp.items().trim().equals("")) {
							if (!rp.type().equals("array")) {
								throw new IllegalArgumentException(
										"请求参数 [ "
												+ rp.name()
												+ " ]存在可选值(items)的时候，请求参数类型(type)的值只能为array");
							}
							Item item = items.get(rp.items().trim());
							if (item != null) { // 可选值
	
								Map<String, Object> i = new HashMap<String, Object>();
								i.put("type", item.type());
								i.put("default", item.defaultValue());
								if (item.type().equals("string")) { // string
									i.put("enum", item.optionalValue());
								} else if (item.type().equals("boolean")) { // boolean
									List<Boolean> bs = new ArrayList<Boolean>();
									for (String v : item.optionalValue()) {
										bs.add(Boolean.parseBoolean(v));
									}
									i.put("enum", bs);
								} else if (item.type().equals("integer")) { // integer
									List<Integer> is = new ArrayList<Integer>();
									for (String v : item.optionalValue()) {
										is.add(Integer.parseInt(v));
									}
									i.put("enum", is);
								} else { // double
									List<Double> ds = new ArrayList<Double>();
									for (String v : item.optionalValue()) {
										ds.add(Double.parseDouble(v));
									}
									i.put("enum", ds);
								}
								parameter.put("items", i);
							}
						}
					}
					parameters.add(parameter);
				}
				p.setParameters(parameters);
			}
		}
		return paths;
	}

	/**
	 * 判断接口的请求Content-Type是否为multipart/form-data。
	 * 
	 * @param service
	 * @return
	 */
	private boolean hasMultipart(API service) {
		for (String consume : service.consumes()) {
			if ("multipart/form-data".equals(consume)) {
				return true;
			}
		}
		return false;
	}

	/**
	 * 解析全部Tag。
	 * 
	 * @return 全部Tag。
	 * @throws Exception
	 */
	private List<Tag> parseTag() throws Exception {
		List<Tag> tags = new ArrayList<Tag>();

		for (Package pk : packages) {
			APITags apiTags = pk.getAnnotation(APITags.class);
			if (apiTags == null) {
				continue;
			}

			APITag[] ts = apiTags.tags();
			for (APITag t : ts) {
				Tag tag = new Tag();
				tag.setName(t.value());
				tag.setDescription(t.description());
				tags.add(tag);
			}
		}
		return tags;
	}

	/**
	 * 解析全部definition。
	 * 
	 * @return 全部definition
	 * @throws Exception
	 */
	private Map<String, Object> parseDefinition() throws Exception {
		Map<String, Object> definitions = new HashMap<String, Object>();

		for (Package pk : packages) {
			APISchemas apiSchemas = pk
					.getAnnotation(APISchemas.class);
			if (apiSchemas == null) {
				continue;
			}
			APISchema[] schemas = apiSchemas.schemas();
			for (APISchema schema : schemas) {
				Map<String, Object> definition = new HashMap<String, Object>();
				definition.put("type", schema.type());
				List<String> required = new ArrayList<String>();
				definition.put("required", required);
				APISchemaPropertie[] props = schema.properties();
				Map<String, Map<String, Object>> properties = new HashMap<String, Map<String, Object>>();
				for (APISchemaPropertie prop : props) {
					Map<String, Object> propertie = new HashMap<String, Object>();
					definition.put("properties", properties);

					propertie.put("type", prop.type());
					propertie.put("format", prop.format());
					propertie.put("description", prop.description());

					if (prop.required()) { // 为必须参数
						required.add(prop.value());
					}

					if (prop.optionalValue().length > 0) { // 可选值
						if (prop.type().equals("string")) { // string
							propertie.put("enum", prop.optionalValue());
						} else if (prop.type().equals("boolean")) { // boolean
							List<Boolean> bs = new ArrayList<Boolean>();
							for (String v : prop.optionalValue()) {
								bs.add(Boolean.parseBoolean(v));
							}
							propertie.put("enum", bs);
						} else if (prop.type().equals("integer")) { // integer
							List<Integer> is = new ArrayList<Integer>();
							for (String v : prop.optionalValue()) {
								is.add(Integer.parseInt(v));
							}
							propertie.put("enum", is);
						} else { // double
							List<Double> ds = new ArrayList<Double>();
							for (String v : prop.optionalValue()) {
								ds.add(Double.parseDouble(v));
							}
							propertie.put("enum", ds);
						}
					}
					properties.put(prop.value(), propertie);
				}
				definitions.put(schema.value(), definition); // 添加新的definition
			}
		}
		return definitions;
	}

	/**
	 * 扫描所有用注解{@link API}修饰了的方法。
	 * 
	 * @return 所有用注解{@link API}修饰了的方法
	 * @throws Exception
	 */
	private List<Method> scanAPIMethod(Class<?> clazz) throws Exception {
		List<Method> api = new ArrayList<Method>();
		APIs apis = clazz.getAnnotation(APIs.class);
		if (apis != null) {
			Method[] methods = clazz.getDeclaredMethods();
			for (Method method : methods) {
				API service = method.getAnnotation(API.class);
				if (service != null) {
					api.add(method);
				}
			}
		}
		return api;
	}
}

File: src/main/kotlin/com/cpjit/swagger4j/APIParser.kt
/*
 * Copyright 2011-2018 CPJIT Group.
 *
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.cpjit.swagger4j

import com.alibaba.fastjson.JSONWriter
import com.cpjit.swagger4j.annotation.*
import com.cpjit.swagger4j.parameter.BodyParameterResolver
import com.cpjit.swagger4j.parameter.ParameterResolver
import com.cpjit.swagger4j.parameter.PrimitiveParameterResolver
import com.cpjit.swagger4j.parameter.SchemaObjectParameterResolver
import com.cpjit.swagger4j.specification.*
import org.apache.commons.lang3.StringUtils
import org.slf4j.LoggerFactory
import org.springframework.core.io.Resource
import org.springframework.core.io.support.PathMatchingResourcePatternResolver
import org.springframework.core.io.support.ResourcePatternResolver
import org.springframework.core.type.classreading.CachingMetadataReaderFactory
import java.io.*
import java.lang.reflect.Method
import java.util.Arrays
import java.util.Properties
import kotlin.collections.ArrayList
import kotlin.collections.HashMap

/**
 * 接口解析器。
 *
 * @author yonghaun
 * @since 1.0.0
 */
class APIParser private constructor(schemesStr: String, private val host: String,
                                    /**
                                     * @return 解析完成后存放JSON数据的文件路径。
                                     */
                                    val file: String, packageToScan: String,
                                    private val basePath: String, description: String, termsOfService: String, title: String, version: String, suffix: String) : APIParseable {
    private val specificationTranslator = SpecificationTranslatorImpl()
    private val resourcePatternResolver = PathMatchingResourcePatternResolver()
    private val metadataReaderFactory = CachingMetadataReaderFactory(this.resourcePatternResolver)
    private val parameterResolvers: List<ParameterResolver>
    private var schemes: Array<String>
    var suffix = ""
    var info: APIDocInfo
    /**
     * @return 待解析接口所在包
     */
    private val packageToScan: List<String>
    private var packages: Collection<Package>

    init {
        parameterResolvers = listOf(
                SchemaObjectParameterResolver(),
                BodyParameterResolver(),
                PrimitiveParameterResolver()
        )
        val schemes = when (StringUtils.isNotBlank(schemesStr)) {
            true -> schemesStr.split(",".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
            false -> arrayOf("http")
        }
        // 扫描class并生成文件所需要的参数
        this.schemes = schemes
        this.packageToScan = Arrays.asList(*packageToScan.split("[;,]".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray())
        try {
            packages = scanClass()
                    .map { it.`package` }
                    .toList()
        } catch (e: Exception) {
            throw IllegalStateException("扫描包信息失败", e)
        }

        this.suffix = suffix
        // API文档信息
        info = APIDocInfo()
        // info.setContact(contact);
        info.description = description
        // info.setLicense(license);
        info.termsOfService = termsOfService
        info.title = title
        info.version = version
    }

    @Throws(Exception::class)
    override fun parse() {
        /* 将结果写入文件 */
        val f = File(file)
        if (LOG.isDebugEnabled) {
            LOG.debug(arrayOf("生成的文件保存在=>", f.toString()).joinToString(""))
        }
        val writer = JSONWriter(OutputStreamWriter(FileOutputStream(f), "utf-8"))
        writer.writeObject(parseAndNotStore())
        writer.flush()
        writer.close()
    }

    @Throws(Exception::class)
    override fun parseAndNotStore(): Any {
        val spec = Specification()
        spec.schemes = schemes;
        spec.host = host
        spec.basePath = basePath

        /* 解析全部item */
        parseItem(spec)
        /* 解析全部tag */
        parseTag(spec)
        /* 解析全部definition */
        parseDefinition(spec)
        /* 解析全部path */
        parsePath(spec)
        return specificationTranslator.translate(spec)
    }

    @Throws(Exception::class)
    private fun parseItem(spec: Specification) {
        val items = HashMap<String, Item>()
        packages.mapNotNull { it.getAnnotation(Items::class.java) }
                .forEach({
                    it.items.forEach {
                        items[it.value] = it
                    }
                })
        spec.items = items
    }


    @Throws(Exception::class)
    private fun parsePath(spec: Specification) {
        scanClass().forEach { clazz ->
            val apis = clazz.getAnnotation(APIs::class.java)
            if (apis?.hide == false) {
                scanAPIMethod(clazz)
                        .forEach {
                            api2Path(it, apis, spec)
                        }
            }
        }
    }

    private fun parseApi(method: Method): Api? {
        val api = Api()
        val apiAnnotation = method.getAnnotation(API::class.java)
        if (apiAnnotation != null) {
            api.parameters = apiAnnotation.parameters
            api.tags = apiAnnotation.tags
            api.consumes = apiAnnotation.consumes
            api.deprecated = apiAnnotation.deprecated
            api.hide = apiAnnotation.hide
            api.description = apiAnnotation.description
            api.method = apiAnnotation.method
            api.operationId = apiAnnotation.operationId
            api.produces = apiAnnotation.produces
            api.summary = apiAnnotation.summary
            api.value = apiAnnotation.value
            return api
        }
        val get = method.getAnnotation(Get::class.java)
        if (get != null) {
            api.parameters = get.parameters
            api.tags = get.tags
            api.consumes = get.consumes
            api.deprecated = get.deprecated
            api.hide = get.hide
            api.description = get.description
            api.method = "GET"
            api.operationId = get.operationId
            api.produces = get.produces
            api.summary = get.summary
            api.value = get.value
            return api
        }
        val post = method.getAnnotation(Post::class.java)
        if (post != null) {
            api.parameters = post.parameters
            api.tags = post.tags
            api.consumes = post.consumes
            api.deprecated = post.deprecated
            api.hide = post.hide
            api.description = post.description
            api.method = "POST"
            api.operationId = post.operationId
            api.produces = post.produces
            api.summary = post.summary
            api.value = post.value
            return api
        }
        val put = method.getAnnotation(Put::class.java)
        if (put != null) {
            api.parameters = put.parameters
            api.tags = put.tags
            api.consumes = put.consumes
            api.deprecated = put.deprecated
            api.hide = put.hide
            api.description = put.description
            api.method = "PUT"
            api.operationId = put.operationId
            api.produces = put.produces
            api.summary = put.summary
            api.value = put.value
            return api
        }
        val delete = method.getAnnotation(Delete::class.java)
        if (delete != null) {
            api.parameters = delete.parameters
            api.tags = delete.tags
            api.consumes = delete.consumes
            api.deprecated = delete.deprecated
            api.hide = delete.hide
            api.description = delete.description
            api.method = "DELETE"
            api.operationId = delete.operationId
            api.produces = delete.produces
            api.summary = delete.summary
            api.value = delete.value
            return api
        }
        return null
    }

    private fun api2Path(method: Method, apis: APIs, spec: Specification) {
        val service = parseApi(method)
        if (service!!.hide) {
            return
        }
        val isMultipart = hasMultipart(service)
        val url = when ("" == service.value) {
            true -> apis.value + suffix
            false -> arrayOf(apis.value, "/", service.value, suffix).joinToString("")
        }
        var path: Path? = spec.paths.firstOrNull { it.url.equals(url) } // get/psot/put/delete
        if (path == null) {
            path = Path(url)
            spec.paths.add(path)
        }
        var operation: Operation? = when (service.method.toLowerCase()) {
            "get" -> path.get
            "put" -> path.put
            "post" -> path.post
            "delete" -> path.delete
            "head" -> path.head
            "patch" -> path.patch
            "options" -> path.options
            else -> null
        }
        if (operation == null) {
            val operationId = when (StringUtils.isNotBlank(service.operationId)) {
                true -> service.operationId
                false -> method.name
            }
            operation = Operation(operationId)
            when (service.method.toLowerCase()) {
                "get" -> path.get = operation
                "put" -> path.put = operation
                "post" -> path.post = operation
                "delete" -> path.delete = operation
                "head" -> path.head = operation
                "patch" -> path.patch = operation
                "options" -> path.options = operation
            }
        }
        if (StringUtils.isNotBlank(service.description)) {
            operation.description = service.description
        } else {
            operation.description = service.summary
        }
        var tags = service.tags
        if (service.tags!!.isEmpty()) {
            var ns = apis.value
            if (ns.startsWith("/")) {
                ns = ns.substring(1)
            }
            tags = arrayOf(ns)
        }
        operation.tags = tags
        operation.summary = service.summary
        if (isMultipart) { // multipart/form-data
            operation.consumes = arrayOf("multipart/form-data")
        } else {
            operation.consumes = service.consumes
        }
        operation.produces = service.produces
        operation.deprecated = service.deprecated
        operation.parameters = parseParameters(url, service, spec)
    }

    /**
     * 解析请求参数。
     */
    private fun parseParameters(url: String, api: Api, spec: Specification): List<Parameter> {
        val parameters: MutableList<Parameter> = mutableListOf()
        for (param in api.parameters) {
            val parameterResolver = parameterResolvers.firstOrNull { it.supportsParameter(api, param) }
            parameterResolver?.resolveParameter(url, api, param, parameters, spec)
        }
        return parameters
    }

    private fun parseOptionalValue(type: String, values: Array<String>): Any {
        return when (type) {
            "string" -> values
            "boolean" -> values.map { java.lang.Boolean.parseBoolean(it) }
            "integer" -> values.map { Integer.parseInt(it) }
            else -> values.map { java.lang.Double.parseDouble(it) }
        }
    }

    /**
     * 判断接口的请求Content-Type是否为multipart/form-data。
     */
    private fun hasMultipart(service: Api): Boolean {
        return service.consumes.any({ "multipart/form-data" == it })
    }

    /**
     * 解析全部Tag。
     *
     * @return 全部Tag。
     */
    @Throws(Exception::class)
    private fun parseTag(spec: Specification) {
        // since1.2.2 先扫描被@APITag标注了的类
        val tags = scanClass()
                .mapNotNull { it.getAnnotation(APITag::class.java) }
                .toMutableSet()
        // 扫描package-info上面的@APITags
        packages.mapNotNull { it.getAnnotation(APITags::class.java) }
                .forEach {
                    tags.addAll(it.tags)
                }
        spec.tags = tags.map {
            val tag = Tag(it.value)
            tag.description = it.description
            tag
        }.toMutableList()

    }

    /**
     * 解析全部definition。
     *
     * @return 全部definition
     */
    @Throws(Exception::class)
    private fun parseDefinition(spec: Specification) {
        val definitions: MutableList<Schema> = mutableListOf()
        for (pk in packages) {
            val apiSchemas = pk.getAnnotation(APISchemas::class.java) ?: continue
            val schemas = apiSchemas.schemas
            for (schema in schemas) {
                val definition = Schema(schema.value)
                definition.type = schema.type
                val required = ArrayList<String>()
                // definition.required = required
                val props = schema.properties
                val properties = mutableListOf<Property>()
                definition.properties = properties;
                for (prop in props) {
                    val property = Property()
                    property.type = prop.type
                    property.format = prop.format
                    property.description = prop.description

                    if (prop.required) { // 为必须参数
                        required.add(prop.value)
                    }
//                    if (prop.optionalValue.isNotEmpty()) { // 可选值
//                        property["enum"] = parseOptionalValue(prop.type, prop.optionalValue)
//                    }
                    properties.add(property)
                }
                definitions.add(definition) // 添加新的definition
            }
        }
        spec.definitions = definitions
    }

    /*
     * 扫描所有用注解{@link API}修饰了的方法。
     *
     * @return 所有用注解{@link API}修饰了的方法
     * @throws Exception
     */
    private fun scanAPIMethod(clazz: Class<*>): List<Method> {
        if (clazz.getAnnotation(APIs::class.java) != null) {
            return clazz.declaredMethods
                    .filter({ this.apiAnnotationedMethod(it) })
        }
        return emptyList()
    }

    private fun apiAnnotationedMethod(method: Method): Boolean {
        return (method.getAnnotation(API::class.java) != null
                || method.getAnnotation(Get::class.java) != null
                || method.getAnnotation(Post::class.java) != null
                || method.getAnnotation(Put::class.java) != null
                || method.getAnnotation(Delete::class.java) != null)
    }


    private fun scanClass(): List<Class<*>> {
        val classes = ArrayList<Class<*>>()
        for (pkg in packageToScan) {
            // 获取包及其子包下面的全部类文件
            val pattern = ResourcePatternResolver.CLASSPATH_ALL_URL_PREFIX + "/" + pkg.replace("\\.".toRegex(), "/") + "/**/*.class"
            var resources: Array<Resource>
            try {
                resources = resourcePatternResolver.getResources(pattern)
            } catch (e: IOException) {
                continue
            }

            for (res in resources) {
                try {
                    val meta = metadataReaderFactory.getMetadataReader(res).classMetadata
                    val clazz = Class.forName(meta.className)
                    if (clazz != null) {
                        classes.add(clazz)
                    }
                } catch (e: IOException) {
                } catch (e: ClassNotFoundException) {
                }

            }
        }
        return classes
    }

    companion object {

        private val LOG = LoggerFactory.getLogger(APIParser::class.java)

        /**
         * 创建一个解析器。
         *
         * @param props properties。
         */
        @Throws(IOException::class)
        fun newInstance(props: Properties): APIParser {
            return APIParser(props.getProperty("schemes"), props.getProperty("apiHost"), props.getProperty("apiFile"), props.getProperty("packageToScan"), props.getProperty("apiBasePath"),
                    props.getProperty("apiDescription"), props.getProperty("termsOfService"), props.getProperty("apiTitle"), props.getProperty("apiVersion"), props.getProperty("suffix"))
        }
    }

}
